File: python/tvm/relax/analysis/dataflow_analysis.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""
Python bindings for the dataflow analysis framework
"""
from enum import Enum
from typing import Any, Callable, List, Tuple
import tvm
from tvm.ir.base import Node
from tvm.relax.expr import Expr, SeqExpr, Function, Var
from . import _ffi_api


class BindingNodeKind(Enum):
    kBinding = 0
    kIfCond = 1
    kIfMerge = 2
    kSeqBody = 3


@tvm._ffi.register_object("relax.analysis.GraphBinding")
class GraphBinding(Node):
    """Representation of a binding in a control flow graph"""

    seq: SeqExpr
    args: List[Var]
    block_idx: int
    binding_idx: int
    kind: BindingNodeKind

    def __init__(
        self,
        seq: SeqExpr,
        args: List[Var],
        block_idx: int,
        binding_idx: int,
        kind: BindingNodeKind,
    ):
        """
        Create a graph binding

        Parameters
        ----------
        seq: SeqExpr
            The SeqExpr that contains the binding

        args: List[Var]
            Arguments taken by the binding (only used for the entry binding:
            these will be the function arguments. Otherwise, this array should be empty.)

        block_idx: int
            The index of the block in the SeqExpr's block list where the binding resides
            (convention: for the SeqExpr body, we will use one past the final block)

        binding_idx: int
            The index of the binding in the binding block corresponding to this binding.

        kind: BindingNodeKind
            The kind of binding. We distinguish between ordinary bindings,
            If conditions, If merges (the var bound to the result of the If node),
            and the body of the SeqExpr.
        """
        return self.__init_handle_by_constructor__(
            _ffi_api.GraphBinding,
            seq,
            args,
            block_idx,
            binding_idx,
            kind,
        )  # type: ignore


@tvm._ffi.register_object("relax.analysis.ControlFlowGraph")
class ControlFlowGraph(Node):
    """Representation of a control flow graph, marking the successors
    and predecessors to all basic blocks"""

    def __init__(
        self, bindings: List[GraphBinding], preds: List[List[int]], succs: List[List[int]]
    ):
        """
        Instantiate a control flow graph

        Parameters
        ----------
        bindings: List[GraphBnding]
            List of bindings in the graph

        preds: List[List[int]]
            The ith member is the list of predecessors to bindings[i] (given as indices in bindings)

        succs: List[List[int]]
            The ith member is the list of successors to bindings[i] (given as indices in bindings)
        """
        if len(bindings) != len(preds) or len(bindings) != len(succs):
            raise ValueError("The lengths of blocks, preds, and succs must all match.")

        return self.__init_handle_by_constructor__(
            _ffi_api.ControlFlowGraph, bindings, preds, succs
        )  # type: ignore


def ExtractCFG(func: Function) -> ControlFlowGraph:
    """
    Given a Relax function, produces the corresponding control flow graph.
    The function is expected to have been normalized.

    Parameters
    ----------
    func: Function
        A Relax function. Must be in normal form.

    Returns
    -------
    graph: ControlFlowGraph
        Control flow graph corresponding to the function.
    """
    return _ffi_api.ExtractCFG(func)  # type: ignore


def GetBindingIndex(
    cfg: ControlFlowGraph, seq: SeqExpr, block_idx: int, binding_idx: int, match_cond: bool = False
) -> int:
    """
    Helper function. Given a control flow graph and a seq expression with a block index and binding index,
    return the index of the corresponding GraphBinding in the CFG

    Parameters
    ----------
    cfg: ControlFlowGraph
        The control flow graph.

    seq: SeqExpr
        The target SeqExpr.

    block_idx: int
        The index of the target block in seq.
        Convention: If the target is `seq.body`, block_idx should be one past the last block
        (i.e., it should be equal to `len(seq.blocks)`).

    binding_idx: int
        The index of the target binding in the target block.

    match_cond: bool
        If true and the target binding in seq is an IfNode, then this function will return
        the binding index corresponding to the If condition.
        If false, then this function will return the binding index corresponding to the If merge.

    Returns
    -------
    idx: int
        The index of the corresponding GraphBindindg in `cfg.bindings`.
    """
    return _ffi_api.GetBindingIndex(cfg, seq, block_idx, binding_idx, match_cond)  # type: ignore


def DataflowAnalysis(
    cfg: ControlFlowGraph,
    init: Any,
    transfer_func: Callable[[GraphBinding, Any], Any],
    merge_func: Callable[[Any, Any], Any],
    forward: bool = True,
) -> Tuple[List[Any], List[Any]]:
    """
    Generic dataflow analysis framework, based on Adrian Sampson's course notes,
    except binding by binding instead of basic block by basic block:
    https://www.cs.cornell.edu/courses/cs6120/2020fa/lesson/4/

    The analysis creates input and output maps (mapping binding indices to a domain),
    sets the initial input and output for each binding to the init value, and then
    performs a traversal of the CFG (BFS in this implementation, since unlike the general case,
    we do not have loops) and uses the transfer and merge function to update the inputs and
    outputs. The analysis can proceed forwards (from binding 0 onwards) or backwards (from the last
    binding back), flipping the roles of the input and output maps in the cases.

    Parameters
    ----------
    cfg: ControlFlowGraph
        The input control flow graph

    init: Any
        The initial value in the analysis domain to which all blocks should be initialized.

    transfer_func: Callable[[GraphBinding, Any], Any]
        Given a binding and the input domain, compute the new output domain.

    merge_func: Callable[[Any, Any], Any]
        When two output domains are fed into a single block (i.e., after an If branch),
        the merge function is used to combine them into a single domain.

    forward: bool
        If true, the analysis proceeds forwards (starting from binding 0 and going onwards).
        If false, the analysis proceeds backwards (starting from the last binding and going back).
        The input and output maps play the opposite roles in forward and backward analyses.
        I.e., in a backward analysis, the "final output" is the input map entry for binding 0
        and the initial input is the output map entry for the last binding.

    Returns
    -------
    ret: Tuple[List[Any], List[Any]]
        A pair of the final input and output maps
    """
    return _ffi_api.DataflowAnalysis(cfg, init, transfer_func, merge_func, forward)  # type: ignore

File: src/relax/analysis/dataflow_analysis.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file tvm/relax/analysis/dataflow_analysis.cc
 * \brief Implementation of functionality in dataflow_analysis.h
 */
#include <tvm/relax/dataflow_analysis.h>
#include <tvm/runtime/memory.h>

#include <queue>

namespace tvm {
namespace relax {

TVM_REGISTER_NODE_TYPE(GraphBindingNode);

GraphBinding::GraphBinding(const SeqExpr& seq, const Array<Var>& args, size_t block_idx,
                           size_t binding_idx, BindingNodeKind kind) {
  ObjectPtr<GraphBindingNode> n = make_object<GraphBindingNode>();
  n->seq = seq;
  n->args = args;
  n->block_idx = block_idx;
  n->binding_idx = binding_idx;
  n->kind = kind;
  data_ = std::move(n);
}

TVM_REGISTER_NODE_TYPE(ControlFlowGraphNode);

ControlFlowGraph::ControlFlowGraph(const Array<GraphBinding>& bindings,
                                   const Array<Array<Integer>>& preds,
                                   const Array<Array<Integer>>& succs) {
  ObjectPtr<ControlFlowGraphNode> n = make_object<ControlFlowGraphNode>();
  n->bindings = bindings;
  n->preds = preds;
  n->succs = succs;
  data_ = std::move(n);
}

// Extracts a basic block and updates the running lists bindings, preds, and succs.
// The return value is the index of the final binding processed in the seq expression
// (useful for processing branches).
size_t ExtractCFGHelper(const SeqExpr& seq, const Array<Var>& args, size_t block_idx,
                        size_t binding_idx, std::vector<size_t> current_preds,
                        std::vector<GraphBinding>* bindings,
                        std::vector<std::vector<size_t>>* preds,
                        std::vector<std::vector<size_t>>* succs) {
  // case 1: We're past the end -> this is the block body (base case)
  if (block_idx == seq->blocks.size()) {
    bindings->push_back(GraphBinding(seq, args, block_idx, 0U, BindingNodeKind::kSeqBody));
    preds->push_back(current_preds);
    // the final binding has no successors
    succs->push_back({});
    return bindings->size() - 1;
  }

  Binding binding = seq->blocks[block_idx]->bindings[binding_idx];
  Expr binding_value = GetBoundValue(binding);

  // case 2: Ordinary binding
  if (!binding_value.as<IfNode>()) {
    bindings->push_back(GraphBinding(seq, args, block_idx, binding_idx, BindingNodeKind::kBinding));
    size_t idx = bindings->size() - 1;
    preds->push_back(current_preds);
    // successor: the next binding (there will always be at least one binding after this,
    // even if it's the seq body)
    succs->push_back({idx + 1});
  } else {
    // case 3: dealing with a branch
    auto if_node = Downcast<If>(binding_value);
    // start with the cond node
    bindings->push_back(GraphBinding(seq, args, block_idx, binding_idx, BindingNodeKind::kIfCond));
    size_t idx = bindings->size() - 1;
    preds->push_back(current_preds);
    // there will be another successor, which we will add after recursing down the branches
    succs->push_back({idx + 1});
    size_t final_true_idx = ExtractCFGHelper(Downcast<SeqExpr>(if_node->true_branch), {}, 0U, 0U,
                                             {idx}, bindings, preds, succs);
    succs->at(idx).push_back(final_true_idx + 1);
    size_t final_false_idx = ExtractCFGHelper(Downcast<SeqExpr>(if_node->false_branch), {}, 0U, 0U,
                                              {idx}, bindings, preds, succs);
    // now create the merge
    bindings->push_back(GraphBinding(seq, {}, block_idx, binding_idx, BindingNodeKind::kIfMerge));
    size_t merge_idx = bindings->size() - 1;
    preds->push_back({final_true_idx, final_false_idx});
    succs->push_back({merge_idx + 1});
    // update the successors of the final true and false indices as well
    succs->at(final_true_idx).push_back(merge_idx);
    succs->at(final_false_idx).push_back(merge_idx);
  }
  // move on to next binding
  size_t next_block_idx = block_idx;
  size_t next_binding_idx = binding_idx + 1;
  if (next_binding_idx >= seq->blocks[block_idx]->bindings.size()) {
    next_block_idx = block_idx + 1;
    next_binding_idx = 0U;
  }
  return ExtractCFGHelper(seq, {}, next_block_idx, next_binding_idx, {bindings->size() - 1},
                          bindings, preds, succs);
}

ControlFlowGraph ExtractCFG(const Function& func) {
  std::vector<GraphBinding> bindings;
  std::vector<std::vector<size_t>> preds;
  std::vector<std::vector<size_t>> succs;
  ExtractCFGHelper(Downcast<SeqExpr>(func->body), func->params, 0U, 0U, {}, &bindings, &preds,
                   &succs);

  Array<Array<Integer>> pred_arr;
  for (auto pred_vec : preds) {
    Array<Integer> pred_ints;
    for (auto idx : pred_vec) {
      pred_ints.push_back(Integer(idx));
    }
    pred_arr.push_back(pred_ints);
  }
  Array<Array<Integer>> succ_arr;
  for (auto succ_vec : succs) {
    Array<Integer> succ_ints;
    for (auto idx : succ_vec) {
      succ_ints.push_back(Integer(idx));
    }
    succ_arr.push_back(succ_ints);
  }
  return ControlFlowGraph(Array<GraphBinding>(bindings), pred_arr, succ_arr);
}

std::pair<Array<ObjectRef>, Array<ObjectRef>> DataflowAnalysis(
    const ControlFlowGraph& cfg, const ObjectRef& init,
    std::function<ObjectRef(const GraphBinding&, const ObjectRef&)> transfer_func,
    std::function<ObjectRef(const ObjectRef&, const ObjectRef&)> merge_func, bool forward) {
  std::vector<ObjectRef> in_map;
  std::vector<ObjectRef> out_map;
  for (size_t i = 0; i < cfg->bindings.size(); i++) {
    in_map.push_back(init);
    out_map.push_back(init);
  }

  // Modification from Adrian Sampson's version:
  // Since there are no loops in our AST, one traversal through the CFG suffices.
  // We will do BFS
  std::queue<size_t> worklist;
  worklist.push((forward) ? 0 : cfg->bindings.size() - 1);
  while (!worklist.empty()) {
    size_t idx = worklist.front();
    worklist.pop();
    Array<Integer> prev = (forward) ? cfg->preds[idx] : cfg->succs[idx];
    Array<Integer> next = (forward) ? cfg->succs[idx] : cfg->preds[idx];
    std::vector<ObjectRef>* results = (forward) ? &out_map : &in_map;
    std::vector<ObjectRef>* inputs = (forward) ? &in_map : &out_map;

    // Cases (for forward analysis):
    // 0 predecessors: The first block in the function
    // 1 predecessor: A branch in an If node (no merge needed)
    // 2 predecessors: The merge block after an If node (merge needed)
    // (Analogous for successors in backward analysis)
    inputs->operator[](idx) = (prev.size() == 0)   ? init
                              : (prev.size() == 1) ? results->at(prev[0].IntValue())
                                                   : merge_func(results->at(prev[0].IntValue()),
                                                                results->at(prev[1].IntValue()));
    results->operator[](idx) = transfer_func(cfg->bindings[idx], inputs->at(idx));

    for (Integer next_idx : next) {
      worklist.push(next_idx.IntValue());
    }
  }

  return {Array<ObjectRef>(in_map), Array<ObjectRef>(out_map)};
}

size_t GetBindingIndex(const ControlFlowGraph& cfg, const SeqExpr& seq, size_t block_idx,
                       size_t binding_idx, bool match_cond) {
  bool is_body = (block_idx == seq->blocks.size());
  bool is_if =
      (!is_body && (GetBoundValue(seq->blocks[block_idx]->bindings[binding_idx]).as<IfNode>()));

  // This is an inefficient linear scan; it could be improved by keeping a map of
  // SeqExprs to indices in the CFG data structure.
  // That should be considered if this function poses performance issues (unlikely).
  for (size_t i = 0; i < cfg->bindings.size(); i++) {
    auto binding = cfg->bindings[i];
    if (binding->seq != seq) {
      continue;
    }
    if (is_body && binding->kind == BindingNodeKind::kSeqBody) {
      return i;
    }
    if (binding->block_idx == block_idx && binding->binding_idx == binding_idx) {
      if (!is_if || (match_cond && binding->kind == BindingNodeKind::kIfCond) ||
          (!match_cond && binding->kind == BindingNodeKind::kIfMerge)) {
        return i;
      }
    }
  }
  CHECK(false) << "Target binding does not appear in the given CFG";
  return cfg->bindings.size();
}

TVM_REGISTER_GLOBAL("relax.analysis.GraphBinding")
    .set_body_typed([](const SeqExpr& seq, const Array<Var>& args, size_t block_idx,
                       size_t binding_idx, int kind) {
      return GraphBinding(seq, args, block_idx, binding_idx, static_cast<BindingNodeKind>(kind));
    });

TVM_REGISTER_GLOBAL("relax.analysis.ControlFlowGraph")
    .set_body_typed([](const Array<GraphBinding>& blocks, const Array<Array<Integer>>& preds,
                       const Array<Array<Integer>>& succs) {
      return ControlFlowGraph(blocks, preds, succs);
    });

TVM_REGISTER_GLOBAL("relax.analysis.ExtractCFG").set_body_typed(ExtractCFG);

TVM_REGISTER_GLOBAL("relax.analysis.DataflowAnalysis")
    .set_body_typed([](const ControlFlowGraph& cfg, const ObjectRef& init, PackedFunc transfer_func,
                       PackedFunc merge_func, bool forward) {
      auto ret = DataflowAnalysis(cfg, init, transfer_func, merge_func, forward);
      return Array<ObjectRef>({ret.first, ret.second});
    });

// need to turn the size_t's into ints in order to cross the C++<->Python boundary
TVM_REGISTER_GLOBAL("relax.analysis.GetBindingIndex")
    .set_body_typed([](const ControlFlowGraph& cfg, const SeqExpr& seq, int block_idx,
                       int binding_idx, bool match_cond) -> int {
      return GetBindingIndex(cfg, seq, block_idx, binding_idx, match_cond);
    });

}  // namespace relax
}  // namespace tvm
